### PR TITLE
New sign tra

### PR DESCRIPTION
**Pull request by HanslettTheDev**
## Summary
This pull request aims to split the sign tra function based on this [comment](https://github.com/PyAr/pyafipws/issues/110#issuecomment-1620976667) from @reingart.
Unit tests will also be added for all the functions to ensure they are properly tested and maintain the standards of the previous general function

## Checklist

- [x] Classes, Variables, function and methods logic ok
- [x] Comments written explaining what the code does

### CODE REVIEW

File: wsaa.py [hunk 1]
```diff
@@ -107,12 +107,53 @@ def create_tra(service=SERVICE, ttl=2400):
...
 
-    if isinstance(tra, str):
-        tra = tra.encode("utf8")
 
+
+def sign_tra_old(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
+    """Legacy method for signing python 2.7 digital transactions on python 2.7 
+    and python 3 versions that don't support cryptography version >39
+    """
     if Binding:
```
Comment (reingart) [inline wsaa.py]: This condition should be removed (and unindent), it already is at top level `sign_tra()` function

File: wsaa.py [hunk 2]
```diff
@@ -188,32 +219,46 @@ def sign_tra(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
...
+
+
+def sign_tra(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
+    "Firmar PKCS#7 el TRA y devolver CMS (recortando los headers SMIME)"
+
+    if isinstance(tra, str):
+        tra = tra.encode("utf8")
+
+    if Binding:
+        if sys.version_info.major == 2:
```
Comment (reingart) [inline wsaa.py]: This also could be needed in Python 3 if using outdated libraries.
The condition should be rewritten to check cryptography version < 39

File: wsaa.py [hunk 3]
```diff
@@ -107,12 +107,53 @@ def create_tra(service=SERVICE, ttl=2400):
...
+            return part.get_payload(decode=False)
+    else:
+        raise RuntimeError("Part not found")
 
-    if isinstance(tra, str):
-        tra = tra.encode("utf8")
 
+
+def sign_tra_old(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
+    """Legacy method for signing python 2.7 digital transactions on python 2.7
```
Comment (reingart) [inline wsaa.py]: Google docstrings style should be used (first line with a summary, then blank, description and args/returns). Please see:

https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings

**Review by reingart — CHANGES_REQUESTED** [review]
Thank you @HanslettTheDev for the contribution!
I made some comments about styling and rewriting some conditions.
Don't forget to add the Unit Tests so this  new functions could get coverage.

**Review by reingart — CHANGES_REQUESTED** [review]
Hi @HanslettTheDev , the tests doesn't have any new UT (all changes are just whitespaces)

At least a `test_sign_tra_openssl` function should be added, that would work in any python / cryptography version.

For the old function version, you could just add a manual tests run evidence (running the python 2 virtualenv and attaching tests results)

TIP: use pytest coverage report:
```
pytest --cov-report term-missing --cov=pyafipws.wsaa tests/test_wsaa.py
````
Then you could see the report for wsaa.py file and confirm that the lines are being covered (exercised by the test):
```
---------- coverage: platform linux, python 3.10.6-final-0 -----------
Name      Stmts   Miss  Cover   Missing
---------------------------------------
wsaa.py     352     89    75%   58-65, 130, 177-199, 203-215, 227, 230-231, 274-276, 294, 323-325, 353-355, 380, 392, 406, 438, 445, 468, 474, 480-481, 485, 491-500, 513-533, 536-540, 552-561, 580-584, 613-614, 627-629, 635, 645
---------------------------------------
TOTAL       352     89    75%
```

**Comment by HanslettTheDev** [comment]
The unit testing functions are with me locally. I was working on a bunch of things so I will push all the commits today

**Comment by HanslettTheDev** [comment]
Manual test evidence of Pytest Test coverage on Python 2.7
image

File: wsaa.py [hunk 4]
```diff
@@ -107,112 +111,154 @@ def create_tra(service=SERVICE, ttl=2400):
...
-            raise RuntimeError("Part not found")
 
+def sign_tra(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
+    "Firmar PKCS#7 el TRA y devolver CMS (recortando los headers SMIME)"
+
+    if isinstance(tra, str):
+        tra = tra.encode("utf8")
+
+    if Binding:
+        if sys.version_info.major == 2 or int(cryptography_version[0:2]) < 39:
```
Comment (reingart) [inline wsaa.py]: Please change this version checking to use split, or it will fail if cryptography isues a release >=100

**Review by reingart — CHANGES_REQUESTED** [review]

File: wsaa.py (hunk 4 above)
Comment (HanslettTheDev) [reply wsaa.py]: done

File: wsaa.py [hunk 5]
```diff
@@ -65,6 +65,10 @@
     from subprocess import Popen, PIPE
     from base64 import b64encode
 
+if "pytest" in sys.modules:
```
Comment (reingart) [inline wsaa.py]: You could import subprocess and base64 always, please do not put conditions for tests in the main code

File: wsaa.py [hunk 6]
```diff
@@ -107,112 +111,154 @@ def create_tra(service=SERVICE, ttl=2400):
...
-            raise RuntimeError("Part not found")
 
+def sign_tra(tra, cert=CERT, privatekey=PRIVATEKEY, passphrase=""):
+    "Firmar PKCS#7 el TRA y devolver CMS (recortando los headers SMIME)"
+
+    if isinstance(tra, str):
+        tra = tra.encode("utf8")
+
+    if Binding:
+        if sys.version_info.major == 2 or int(cryptography_version.split(".")[0]) < 39:
```
Comment (reingart) [inline wsaa.py]: If the last supported version for old logic was 39, this would be `<=` right?

**Review by reingart — CHANGES_REQUESTED** [review]

File: wsaa.py (hunk 5 above)
Comment (HanslettTheDev) [reply wsaa.py]: Got it, I will remove that line. I added that line of code as the tests were failing due to subprocess and base64 not imported at run time due to the try and except statement

File: wsaa.py (hunk 6 above)
Comment (HanslettTheDev) [reply wsaa.py]: oh I modified the code and forgot to add the =
It has been modified now

**Comment by HanslettTheDev** [comment]
Coverage Report for both Python 2.7 and 3+

[htmlcov-python3+.zip](https://github.com/PyAr/pyafipws/files/12413270/htmlcov-python3%2B.zip)
[htmlcov-python27.zip](https://github.com/PyAr/pyafipws/files/12413271/htmlcov-python27.zip)

**Comment by reingart** [comment]
Thanks for the coverage! the python2 zip seems to miss the wsaa.py html, could you re-upload or generate a new report?

**Comment by HanslettTheDev** [comment]
Reuploaded the test coverage report for python2.7
[htmlcov-python27.zip](https://github.com/PyAr/pyafipws/files/12415510/htmlcov-python27.zip)

**Review by reingart — APPROVED** [review]

**Pull request merged**